src/data/list/of_fn.lean

```diff
@@ -107,6 +107,21 @@ begin
     simp_rw [fin.cast_succ_fin_succ], }
 end
 
+@[simp] lemma of_fn_eq_nil_iff {n : ℕ} {f : fin n → α} :
+  of_fn f = [] ↔ n = 0 :=
+by cases n; simp only [of_fn_zero, of_fn_succ, eq_self_iff_true, nat.succ_ne_zero]
+
+@[simp] lemma last_of_fn {n : ℕ} (f : fin n → α) (h : of_fn f ≠ [])
+  (hn : n - 1 < n := lt_of_not_le (mt (λ hn,
+    of_fn_eq_nil_iff.mpr (nat.pred_eq_self_iff.mp (le_antisymm (nat.pred_le _) hn))) h)) :
```

Comment:
Is there a precedent for using a default argument like this? It worries me expecially with `simp`. Maybe better to split `hn` into a separate lemma and inline it in the type?

Reply:
Some examples, from various users, from `list/basic`:
```
-- Floris
lemma nth_le_tail (l : list α) (i) (h : i < l.tail.length)
  (h' : i + 1 < l.length := by simpa [←lt_tsub_iff_right] using h) :

-- EricR
lemma nth_le_cons {l : list α} {a : α} {n} (hl) :
  (a :: l).nth_le n hl = if hn : n = 0 then a else l.nth_le (n - 1) (nth_le_cons_aux hn hl) := -- this inlines the proof, which I think is less flexible

-- ChrisH
@[simp] lemma nth_le_update_nth_of_ne {l : list α} {i j : ℕ} (h : i ≠ j) (a : α)
  (hj : j < (l.update_nth i a).length) :
  (l.update_nth i a).nth_le j hj = l.nth_le j (by simpa using hj) :=

-- me
lemma nth_le_enum_from (l : list α) (n i : ℕ)
  (hi' : i < (l.enum_from n).length)
  (hi : i < l.length := by simpa [length_enum_from] using hi') :
```

Reply:
That is, I tried it with `simp [h]`, and that didn't solve it. So I see your point about inlining it. But just untagging it might be fine as it is.

Reply:
```suggestion
  (hn : n - 1 < n := nat.pred_lt $ of_fn_eq_nil_iff.not.mp h) :
```